### PR TITLE
Migrate `mir_transform` to translatable diagnostics

### PR DESCRIPTION
**Pull request by clubby789**
cc #100717

### CODE REVIEW

File: compiler/rustc_mir_transform/src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+pub(crate) struct RequiresUnsafe {
+    pub span: Span,
+    pub details: RequiresUnsafeDetail,
+    pub enclosing: Option<Span>,
+    pub op_in_unsafe_fn_allowed: bool,
+}
+
+impl<'sess, G: EmissionGuarantee> IntoDiagnostic<'sess, G> for RequiresUnsafe {
+    #[track_caller]
+    fn into_diagnostic(self, handler: &'sess Handler) -> DiagnosticBuilder<'sess, G> {
```
Comment (Nilstrieb) [inline compiler/rustc_mir_transform/src/errors.rs:55]: Why does the derive not work here?

File: compiler/rustc_mir_transform/src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    CallToFunctionWith(Span),
+}
+
+impl RequiresUnsafeDetail {
+    pub fn from_details(details: UnsafetyViolationDetails) -> impl Fn(Span) -> Self {
+        use RequiresUnsafeDetail as B;
+        use UnsafetyViolationDetails as A;
+        match details {
+            A::CallToUnsafeFunction => B::CallToUnsafeFunction,
+            A::UseOfInlineAssembly => B::UseOfInlineAssembly,
```
Comment (Nilstrieb) [inline compiler/rustc_mir_transform/src/errors.rs:93]: Would it make more sense to turn `RequiresUnsafeDetail` into a struct containing the span and unsafety violation? That should remove this redundant looking function.

File: compiler/rustc_mir_transform/src/errors.rs (hunk 1 above)
Comment (clubby789) [reply compiler/rustc_mir_transform/src/errors.rs:55]: We have to eagerly translate another message to be used in the 'requires unsafe' main diagnostic

Comment (clubby789) [reply compiler/rustc_mir_transform/src/errors.rs:56]: I guess I could also duplicate the message and exhaustively list the full message for each case in the .ftl if there's no cleaner way

Comment (oli-obk) [reply compiler/rustc_mir_transform/src/errors.rs:56]: Please add that explanation (and the mentioned undesirable alternative) as a comment to the manual impl

**Pull request merged**